### PR TITLE
refactor(server): use `Deno.serve()`

### PR DESCRIPTION
**Pull request by kwaa**
## Description

A refactoring of Lume Server that uses the new `Deno.serve()` to simplify structure and improve performance.

> I'm not quite sure how to write a changelog for v2 at the moment, so I left it out.

```diff
import Server from "https:/deno.land/x/lume/core/server.ts";

const server = new Server({
  port: 8000,
  root: `${Deno.cwd()}/_site`,
});

- await server.start();
+ server.start();
+ // optional: Deno.serve({ port: 8000 }, server.handle)

console.log("Listening on http://localhost:8000");
```

It's worth noting that Middleware's `ConnInfo` has changed to `Deno.ServeHandlerInfo`.

### Check List

- [x] Have you read the
      [CODE OF CONDUCT](https://github.com/lumeland/lume/blob/master/CODE_OF_CONDUCT.md)
- [x] Have you read the document
      [CONTRIBUTING](https://github.com/lumeland/lume/blob/master/CONTRIBUTING.md)
  - [x] One pull request per feature. If you want to do more than one thing,
        send multiple pull request.
  - [ ] Write tests.
  - [x] Run deno `fmt` to fix the code format before commit.
  - [ ] Document any change in the `CHANGELOG.md`.

### CODE REVIEW

File: core/server.ts [hunk 1]
```diff
@@ -1,32 +1,27 @@
...
 import type { Event, EventListener, EventOptions } from "../core.ts";
-import type { ConnInfo } from "../deps/http.ts";
 
 /** The options to configure the local server */
 export interface Options {
   /** The root path */
   root: string;
-
-  /** The port to listen on */
-  port: number;
```
Comment (oscarotero) [inline core/server.ts:17]: I think server options should keep as they are.

File: core/server.ts [hunk 2]
```diff
@@ -78,23 +73,15 @@ export default class Server {
...
   /** Start the server */
-  async start() {
-    const { port } = this.options;
-
-    this.#server = new HttpServer({
-      port: port,
-      handler: (request, connInfo) => this.handle(request, connInfo),
-    });
-
+  start(options: Deno.ServeOptions = { port: 8000 }) {
```
Comment (oscarotero) [inline core/server.ts:76]: The `start` shouldn't  have any option.

**Review by oscarotero — CHANGES_REQUESTED** [review]
Thanks for this!
Everything looks fine, but I'd like to keep all options in the Options object passed to the constructor, instead of split some options in the constructions and other in the `start()` function.

**Comment by kwaa** [comment]
> Thanks for this! Everything looks fine, but I'd like to keep all options in the Options object passed to the constructor, instead of split some options in the constructions and other in the `start()` function.

In that case, maybe it should be passed as a sub-option?

```ts
/** The options to configure the local server */
export interface Options {
  /** The root path */
  root: string;
  serve?: Deno.ServeOptions
}
```

File: core/server.ts (hunk 2 above)
Comment (oscarotero) [reply core/server.ts:76]: And due this class has events, probably it should handle the `Deno.Server` events automatically. I mean:

```js
options.onError = (error) => this.dispatchEvent({ type: "error", error });
options.onListen = () => this.dispatchEvent({ type: "start" });
```

Comment (kwaa) [reply core/server.ts:76]: > And due this class has events, probably it should handle the `Deno.Server` events automatically. I mean:
> 
> ```js
> options.onError = (error) => this.dispatchEvent({ type: "error", error });
> options.onListen = () => this.dispatchEvent({ type: "start" });
> ```

It looks like `Deno.ServeOptions`'s `onError` needs to return a `Response`.

```ts
interface ServeOptions {
  /** The handler to invoke when route handlers throw an error. */
  onError?: (error: unknown) => Response | Promise<Response>;
  /** The callback which is called when the server starts listening. */
  onListen?: (params: { hostname: string; port: number }) => void;
}
```

**Comment by oscarotero** [comment]
@kwaa This makes more sense to me:

```js
const server = new Server({
  root: `${Deno.cwd()}/_site`,
  port: 8000, // Deno.ServeOptions
  reusePort: false, // Deno.ServeOptions
  hostname: "localhost", // Deno.ServeOptions
});

server.start(signal); // Probably this makes more sense on start the server.

// Other options like onError and onListen should be handled automatically by the current event system.
```

File: core/server.ts (hunk 1 above)
Comment (kwaa) [reply core/server.ts:17]: `Options` now extends `Deno.ServeOptions`.

File: core/server.ts (hunk 2 above)
Comment (oscarotero) [reply core/server.ts]: Okay, in this case, onError can be an option in the constructor. Something like `errorResponse: (error) => Response`

Comment (kwaa) [reply core/server.ts]: So onError executes this.dispatchEvent and returns a this.errorResponse?  Is there a suitable default value?

Comment (oscarotero) [reply core/server.ts]: I don't know. Maybe no default value for now.

**Review by oscarotero — APPROVED** [review]

**Comment by oscarotero** [comment]
Awesome work. Thank you!

**Pull request merged**